LFS-698: Restructure the patient chart

veronikaslc opened the pull request:
[LFS-698](https://phenotips.atlassian.net/browse/LFS-698): Restructure the patient chart
[LFS-715 / LFS-717](https://phenotips.atlassian.net/browse/LFS-717): Subject chart displays questionnaire description
[LFS-736](https://phenotips.atlassian.net/browse/LFS-736): The (+) button should be at the bottom of the pages (Partial implementation: Subject chart, Subjects page, Forms page. Dashboard is unchanged and will be covered by the dashboard redesign #386)
[LFS-594](https://phenotips.atlassian.net/browse/LFS-594): Subject page is not updated after successful deletion of its forms

Commit:
LFS-698: Restructure the patient chart
WIP

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -291,50 +299,56 @@ function SubjectMember (props) {
+	                      title={
+	                        <React.Fragment>
+	                          <Button size={buttonSize} className={classes.subjectFormHeaderButton}>
+	                            {moment(entry["jcr:created"]).format("yyyy-MM-dd")}
```
I think `dd` is for day of the week, I see `2020-11-WE`. Maybe `DD` is for day of the month in `moment.js`?
```suggestion
	                            {moment(entry["jcr:created"]).format("yyyy-MM-DD")}
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -291,50 +299,56 @@ function SubjectMember (props) {
+        { subjectGroups &&
+          Object.keys(subjectGroups).map( (questionnaireName, j) => {
+            return(<>
+	          <Typography variant="h4" className={classes.subjectQuestionnaireSubheader}>
```
I wonder if this should "shrink" with the level, starting at `h4` for the topmost subject (e.g. Patient on a patient chart), `h5` for 1st level children (e.g. Tumors), `h6` for 2nd+ level children (e.g. Tumor regions and everything that might have Tumor region as an ancestor).

Currently there's a bit of mismatch between the "Tumor region <id>" subtitle and its "Somatic variants" form group, but it may start looking more messy if we decrease the size.

Let's revisit this thought after everything else is finalized and ask for feedback from @sashaandjic on it, too.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -224,11 +225,18 @@ function SubjectMember (props) {
+    let groups = {};
+    json.rows.map( (entry, i) => {
+        groups[entry.questionnaire["@name"]] = groups[entry.questionnaire["@name"]] || [];
+        groups[entry.questionnaire["@name"]].push(entry);
```
This can be simplified to avoid self assignment `groups[entry.questionnaire["@name"]] = groups[entry.questionnaire["@name"]]`:
```suggestion
        groups[entry.questionnaire["@name"]]?.push(entry) || (groups[entry.questionnaire["@name"]] = [entry]);
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -291,50 +299,56 @@ function SubjectMember (props) {
+          Object.keys(subjectGroups).map( (questionnaireName, j) => {
+            return(<>
+	          <Typography variant="h4" className={classes.subjectQuestionnaireSubheader}>
+		        {questionnaireName}
```
The questionnaire `title` should be displayed here, with fallback to `@name`, instead of its `@name` directly.

(Note: after [LFS-714](https://phenotips.atlassian.net/browse/LFS-714) _Remove "Data" from all LFS Questionnaire titles_ the title will happen to be identical to the name for LFS, and will match what was depicted in the mockup. For projects other than LFS, questionnaire titles are usually more elaborate).

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -291,50 +299,56 @@ function SubjectMember (props) {
-          ) : ""
+        { subjectGroups &&
+          Object.keys(subjectGroups).map( (questionnaireName, j) => {
+            return(<>
```
Each `React.Fragment` needs a `key`.

Commit:
LFS-698: Restructure the patient chart
Addressed test comments

sashaandjic commented:
Here's the current look:
[LFS-698.pdf](https://github.com/ccmbioinfo/lfs/files/5569431/LFS-698.pdf)

The trash cans should be bottom aligned.to the corresponding title.
I'm also now thinking if we should move the trash cans next to the dates? WDYT?

sashaandjic commented:
I got this console error when I clicked on the + button and chose Radiotherapy form
![image](https://user-images.githubusercontent.com/23243244/99719497-5aa3e800-2a7a-11eb-9572-0d4d22ba43fb.png)

Commit:
LFS-698: Restructure the patient chart
Forms corresponding to each questionnaire should be listed in a table

marta- reviewed (changes requested):
A couple of issues:
* The questionnaire titles seem to be missing in the latest version of the branch, all I see is the tables (but forms are correctly split in separate tables by questionnaire). They could be displayed as the MaterialTable title.
* The tables would look better if the cell contents were vertically aligned. Also the summary card in the Summary column shouldn't have any padding, and the margin of the status chips should be switched from left to right in the new layout.
* The search associated with each table doesn't seem to find anything atm, and the repetitive search input does not look good on the screen. We don't expect that many entries per questionnaire to justify this searching functionality, so it should be removed.

Additionally, LFS-594 (form entry doesn't disappear from subject char after being successfully deleted) persists in the new layout as well. Please fix it in the same branch.

Commit:
LFS-698: Restructure the patient chart
Addressing codereview comments

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -365,6 +365,9 @@ const questionnaireStyle = theme => ({
     hierarchyEditButton: {
         marginLeft: theme.spacing(1)
     },
+    subjectQuestionnaireSubheader: {
```
This is no longer used anywhere

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -291,50 +301,73 @@ function SubjectMember (props) {
+                      padding: '0',
+                      paddingTop: '8px'
+                    },
+                    render: rowData => <CardContent>
```
FormData doesn't need to be wrapped in a CardContent anymore since it's no longer displayed in a Card. This should also help fix the padding difference.

Also the import of Card, CardHeader, and CardContent should be removed.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -291,50 +301,73 @@ function SubjectMember (props) {
-          ) : ""
+        { subjectGroups &&
+          Object.keys(subjectGroups).map( (questionnaireName, j) => {
+            return(<React.Fragment key={questionnaireName}>
```
The tables displayed are too close together and could use some vertical spacing. Wrapping them in their own grid item might help with that.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -291,50 +301,73 @@ function SubjectMember (props) {
+                }}
+                columns={[
+                  { title: 'Created',
+                    cellStyle: {
```
The custom cell padding is unnecessary, especially after removing the CardContent wrapper for FormData.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -291,50 +301,73 @@ function SubjectMember (props) {
-          </Grid>
-          ) : ""
+        { subjectGroups &&
+          Object.keys(subjectGroups).map( (questionnaireName, j) => {
```
Technically it's the questionnaire title, so we should name the variable accordingly
```suggestion
          Object.keys(subjectGroups).map( (questionnaireTitle, j) => {
```

sashaandjic commented:
Here is the new look:
![image](https://user-images.githubusercontent.com/23243244/100396116-d4efe180-3011-11eb-9b72-f6d12763b619.png)

sashaandjic commented:
Would this look better:
![image](https://user-images.githubusercontent.com/23243244/100396145-f4870a00-3011-11eb-815e-9543c6ab767e.png)

sashaandjic commented:
Or even simpler:
![image](https://user-images.githubusercontent.com/23243244/100396779-724c1500-3014-11eb-98f6-261853572f55.png)

sashaandjic commented:
Without # of rows
![image](https://user-images.githubusercontent.com/23243244/100397444-b7257b00-3017-11eb-8a52-93e2a0e78374.png)

Commit:
LFS-594: Subject page is not updated after successful deletion of its form(s)

Commit:
LFS-698: Restructure the patient chart
Addressed codereview comments for better look

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -291,51 +298,62 @@ function SubjectMember (props) {
-            })}
-          </Grid>
-          ) : ""
+        <Grid>{ subjectGroups &&
```
```suggestion
        <Grid container spacing={8}>{ subjectGroups &&
```

veronikaslc committed:
LFS-698: Restructure the patient chart

Added tables spacing

Co-authored-by: Marta Girdea <marta.girdea@gmail.com>

marta- commented:
I'm ok with making the following UI changes:
 1. Remove pencil button (we can add it back later after read-only mode is implemented)
 2. Remove table column labels (Created, Status, Summary)
 3. Shrink the Created column
 4. Remove the bottom border (under pagination)
 5. Change pageSize to 10
 6. Only display the pagination when there are more than pageSize entries to display
 
 Looks like 2 and 6 can easily be done via component overriding https://material-table.com/#/docs/features/component-overriding , and the same mechanism can be used to implement LFS-717 - displaying the questionnaire description.
 
 Other fixes:
 * The actions implementation is incorrect: setting the button as the action icon will generate a button within a button and produce a console error. Since our buttons are already implemented  separately and don't need to rely on MaterialTable's action generation mechanism, I suggest rendering the button in a "normal" column and getting rid of the MaterialTable actions.
 
 **Not** to be done:
 * Table indentation - too much white space on the left
 * Moving the (+) button and aligning the delete buttons - this is out of scope and will be done in https://phenotips.atlassian.net/browse/LFS-736 .

veronikaslc commented:
@marta- May I suggest to use one table with questionnaire filter instead of a separate tale per each questionnaire?

marta- commented:
W

> @marta- May I suggest to use one table with questionnaire filter instead of a separate tables per each questionnaire?

Why?

veronikaslc commented:
> > @marta- May I suggest to use one table with questionnaire filter instead of a separate tables per each questionnaire?
> 
> Why?

LiveTable would be the most appropriate table component to use there since forms query uses pagination servlet. Since we are implementing LFS-576, we may use query filter on the patient chart which makes the look more compact and consistent with the rest app pages.

marta- commented:
Consistency should not get in the way of usability. The patient chart should have more structure. One of the goals of this task was to allow the user to have a good overview of what kind of data has been recorded for a patient, and that was the reason behind grouping the forms by questionnaire and making the questionnaire titles prominent so they can be noticed at a glance. Listing all entries in a table does not provide that kind of overview, filtering by questionnaire requires several extra steps, and with a filtered table we can not display more than one questionnaire at a time.

The questionnaire filter looks great, please make a separate PR for it.

veronikaslc commented:
I can separate the questionnaire filter PR but I will have to rebase this PR on it. Using LiveTable instead of Material Table is unavoidable in the context of this PR since pagination query only returns first N results. There is no way to have separate tables per questionnaire and use this query dynamically without being able to filter by the questionnaire. We can query all Forms setting limit to 1000 but that could affect performance taking into account that deep.json if fetched for each Form multiple times due to nested SubjectContext re-rendering.

marta- commented:
Realistically, none of our current projects are likely to have more than 100 forms per subject. If at any time in the future we will have this issue, we should revisit the design. However, for now, we should focus on making the current data easy to navigate for our current users.

The reason I asked for the questionnaire filter feature to be put in a separate PR is that it is likely to be merged faster than 698, and it will be useful to complete other developments such as the new user dashboard. Please do rebase on top of it until it gets merged into dev.

The UI comments still stand regardless of the widget used. We need to display:
* **Questionnaire title**
* _Questionnaire description, if any_
* List of forms corresponding to that questionnaire, for which we must specify date of entry, status if any, excerpt, and delete button (without any table headers, since each item is self-explanatory)
* Controls to display more entries (only if more entries are available)

Commit:
LFS-698: Restructure the patient chart
Addressing codereview comments

sashaandjic commented:
Tumor chart problems:
- patient link is not working
- tumor regions are overlapping
- the form data needs to be top aligned, as well as the trach bins
![image](https://user-images.githubusercontent.com/23243244/100702072-eff28680-336e-11eb-8873-13fd4ea24da3.png)

Patient chart look:
[LFS-698.pdf](https://github.com/ccmbioinfo/lfs/files/5620505/LFS-698.pdf)

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -204,12 +212,14 @@ function SubjectContainer(props) {
   let [tableData, setTableData] = useState();
   let [subjectGroups, setSubjectGroups] = useState();
+  const location = useLocation();
+  const history = useHistory();
```
`history` doesn't appear to be used anywhere.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -456,7 +466,7 @@ Subject.propTypes = {
 
 Subject.defaultProps = {
-  maxDisplayed: 4
+  maxDisplayed: 10
```
Increasing `maxDisplayed` will generate form excerpts that are too large (with too many fields) that are hard to follow in a table.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -308,16 +308,22 @@ function SubjectMember (props) {
-                  rowStyle: {
-                    verticalAlign: 'top',
-                  }
+                  pageSize: maxDisplayed,
```
`maxDisplayed` was intended for specifying how many form fields to display in the form excerpt.
We shouldn't use the same parameter for two different purposes. Please introduce another variable for table page size.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -18,7 +18,7 @@
 
 import React, { useState } from "react";
-import { Link } from 'react-router-dom';
+import { Link, useLocation, useHistory } from 'react-router-dom';
```
Changes related to the parent chart not loading should have been committed separately and labeled with the issue I'd LFS-749.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -265,28 +275,27 @@ function SubjectMember (props) {
+    if (location.pathname != "/content.html" + parentPath){
+      update && update(location.pathname.replace("/content.html/Subjects/", ""));
+    }
+  }, [location]);
```
This works, but to me it looks quite hackish and brittle. We should work with the `@name` directly instead of deconstructing the `@path`. You correctly identified that the state wasn't passed back to the parent properly which caused the root component `Subject` not to render, but there's an easier way to ensure that, without `useEffect`. I implemented it in draft PR #374, which is inspired by your changes. It updates the state when clicking on the link rather than via effect when the location changes.

EDIT: I was wrong, the easier way without `useEffect` doesn't work properly. But there is a less intricate way to `useEffect` for solving the issue.

marta- replied:
The solution with state update when clicking on the links was incomplete and did not cover re-rendering when using the browser's back button. I updated that PR to re-render via `useEffect` on location change like you did. The differences are:
* `useEffect` is called in the top component, so we avoid propagating the state up from a deeply nested component
* the extraction of the new id is a bit more robust and uses the same logic as in `Subjects.jsx` with no reliance on `content.html`.

sashaandjic commented:
Here is the current look of one massive patient chart.
![LFS-698](https://user-images.githubusercontent.com/23243244/100901687-19f09980-3492-11eb-8490-2e708f3d4676.png)
I think we still have some alignments to do.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -300,53 +319,79 @@ function SubjectMember (props) {
+                                       </React.Fragment> },
+                  { title: 'Summary',
+                    cellStyle: {
+                      padding: '12px 0 5px 0',
```
These padding values look quite arbitrary. I assume they are like this to ensure alignment with the delete button. I would suggest playing with minHeight instead of setting a 5px bottom padding. Also, a `DeleteButton` with `size="small"` might fit better in the table and be easier to align.

veronikaslc replied:
Delete button size did not help with alignment unfortunately.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -300,53 +319,79 @@ function SubjectMember (props) {
+          Object.keys(subjectGroups).map( (questionnaireTitle, j) => {
+            return(<Grid item key={questionnaireTitle}>
+              <MaterialTable
+                title={questionnaireTitle}
```
The space at the left of the title (I believe it is a paddingLeft of 24px) needs to be removed to make the table look less messy.

Commit:
LFS-698: Restructure the patient chart
Addressing codereview comments

sashaandjic commented:
one minor alignment is needed on Tumor chart
![image](https://user-images.githubusercontent.com/23243244/101063138-7a4e0c80-3560-11eb-8c21-da54fd87101a.png)

sashaandjic commented:
here is the look of patient chart:
![image](https://user-images.githubusercontent.com/23243244/101063326-abc6d800-3560-11eb-934e-87b2b2a1aec3.png)

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -453,7 +500,8 @@ Subject.propTypes = {
 
 Subject.defaultProps = {
-  maxDisplayed: 4
+  maxDisplayed: 10,
```
```suggestion
  maxDisplayed: 4,
```

sashaandjic commented:
When running in cardiac_rehab mode I get this on the dashboard
![image](https://user-images.githubusercontent.com/23243244/101066004-dcf4d780-3563-11eb-805b-98fb00f105b7.png)

marta- commented:
> When running in cardiac_rehab mode I get this on the dashboard

There's no cardiac_rehab mode in this branch, dev would need to be merged in locally

veronikaslc committed:
LFS-698: Restructure the patient chart

Setting maxDisplayed back to 4

Co-authored-by: Marta Girdea <marta.girdea@gmail.com>

Commit:
LFS-698: Restructure the patient chart
Fixing allignment

sashaandjic commented:
Chart looks really neat for cardiac rehab mode, however their questions and some answers are quite long. I think the rows should be widen to avoid word wrapping (easier to read). One column rather than 2 seems to provide better overview of the patien chart.
![CR-chart](https://user-images.githubusercontent.com/23243244/101210654-db024580-3643-11eb-8afc-b3a764f1c967.png)

sashaandjic commented:
LFS Patient chart
![LFS-Patient chart](https://user-images.githubusercontent.com/23243244/101232796-d8bedc00-3681-11eb-9a5c-55ccaf580065.png)

sashaandjic commented:
Cardiac patient chart
![image](https://user-images.githubusercontent.com/23243244/101235637-0282fd80-3698-11eb-8b11-7a323507d34b.png)

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -370,6 +365,11 @@ const questionnaireStyle = theme => ({
     },
+    patientChartTableTitle: {
+        left: 0,
+        position: "absolute",
```
Absolute positioning should be used with caution when we deal with text of variable lengths. We have some questionnaire titles in cardiac rehab that will easily overflow.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -288,25 +295,23 @@ function SubjectMember (props) {
-  return (
+  return ( data &&
     <Grid item xs={12}>
       <Grid item>
```
I just noticed there was an item within an item here...

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -315,53 +320,81 @@ function SubjectMember (props) {
-          </Grid>
-          ) : ""
+      { subjectGroups && Object.keys(subjectGroups).length > 0 && <Grid item>
+        <Grid container spacing={8}> {
```
These should be on one column:
```suggestion
        <Grid container direction="column" spacing={8}> {
```

marta- committed:
LFS-698: Restructure the patient chart
* Displayed questionnaire tables on one column
* Changed how questionnaire title is displayed to protect agains layout issues with long titles
* Simplified markup
* Some spacing improvements

marta- committed:
LFS-715 / LFS-717: Subject chart displays questionnaire description

marta- committed:
LFS-736: The (+) button should be at the bottom of the pages
Moved the button on the Subject chart

marta- commented:
Many of the UI comments should be addressed in the last 3 commits

marta- committed:
LFS-736: The (+) button should be at the bottom of the pages
Moved the button on the Subjects page and the Forms page.
Dashboard is still unchanged pending the redesign (LFS159).

marta- committed:
LFS-698: Restructure the patient chart
Fixed the spacing for child subjects without forms

marta- committed:
LFS-698: Restructure the patient chart
* Fixed bug introduce in previous commits that made subject hierarchy links unclickable
* Added avatars and links to the child subject titles
* Made page header sticky when scrolling down, similar behavior to form page headers

marta- committed:
LFS-698: Restructure the patient chart
LFS-594: Subject page is not updated after successful deletion of its forms
Also ensured the chart gets updated when deleting a child subject

marta- committed:
LFS-698: Restructure the patient chart
Fixed console warning caused by invalid property value

marta- committed:
LFS-698: Restructure the patient chart
Minor ui/color fixes:
* made child subject avatars green to be consistent with the rest other subject avatars (quick search, subject selector)
* removed warning coloring from delete buttons

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -280,89 +315,119 @@ function SubjectMember (props) {
   }
+
   // change styling based on 'level'
   let buttonSize = "large";
```
unused

OrdiNeu reviewed (approved):
Works and looks good.

OrdiNeu committed:
LFS-698: Restructure the patient chart

Remove unused variable

marta- committed:
Merge branch 'dev' into LFS-698

sashaandjic reviewed (approved):
tested/all good

sashaandjic commented:
I just realized the tumor regions would look better if sorted. Not sure if depicted scenario is realistic.
![Patient chart](https://user-images.githubusercontent.com/23243244/101795978-79742800-3ad6-11eb-8566-6ad8b8c36497.png)

marta- committed:
LFS-698: Restructure the patient chart

Display child subjects sorted chronologically by their creation date

sashaandjic commented:
re-tested - tumor and tumor regions are sorted chronologically now - APPROVED!
![image](https://user-images.githubusercontent.com/23243244/101806208-b1cd3380-3ae1-11eb-91b9-12ec92fa3f87.png)

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx:
```diff
@@ -25,19 +25,19 @@ import moment from "moment";
   Button,
 } from "@material-ui/core";
 import DeleteButton from "../dataHomepage/DeleteButton.jsx";
+import EditButton from "../dataHomepage/EditButton.jsx";
```
Not used, should be removed.

```suggestion
```

sdumitriu committed:
[cleanup] Removed unused import
